ui/src/components/FileTreeView/FileTreeView.types.ts

```diff
@@ -23,3 +23,69 @@ export type TreeState = {
   openedFile: NodeAbsoluteLocation;
   openFile: (loc: NodeAbsoluteLocation) => void;
 };
+
+export function getAllDirectories(prefix: string[], files: FileTreeNode[]): string[][] {
+  const dirs: string[][] = [];
+  files
+    .filter(isDirectory)
+    .map(file => file as DirNode)
+    .forEach(dir => {
+      const newPrefix = [...prefix, dir.name];
+      dirs.push(newPrefix);
+      dirs.push(...getAllDirectories(newPrefix, dir.content));
+    });
+  return dirs;
+}
+
+export function getMainScala(files: FileTree): string[] {
```

Comment:
In the `FileTreeView` we have access to the `groupId`, maybe, instead of the search, we could split it and concatenate absolute path from it?

Reply:
Good idea. It simplifies the calculations and `gatMainScala` (and methods it depends on) can be removed.